syntax = "proto3";

option go_package = "github.com/numaproj/numaflow-go/pkg/apis/proto/mapbatch/v1";

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

package mapbatch.v1;

service MapBatch {
  // MapBatchFn applies a function to mulitple elements at once
  rpc MapBatchFn(MapBatchRequest) returns (MapBatchResponse);

  // IsReady is the heartbeat endpoint for gRPC.
  rpc IsReady(google.protobuf.Empty) returns (ReadyResponse);
}

/**
 * MapRequest represents a request element.
 */
message MapRequest {
  repeated string keys = 1;
  bytes value = 2;
  google.protobuf.Timestamp event_time = 3;
  google.protobuf.Timestamp watermark = 4;
  map<string, string> headers = 5;
}

message MapBatchRequest {
  repeated MapRequest messages = 1;
}

/**
 * MapBatchResponse represents a response element.
 */
message MapBatchResponse {
  message Result {
    repeated string keys = 1;
    bytes value = 2;
    repeated string tags = 3;
  }
  repeated Result results = 1;
}

/**
 * ReadyResponse is the health check result.
 */
message ReadyResponse {
  bool ready = 1;
}